File: Platforms/Android/platformchannels/src/main/java/com/microsoft/dotnet/platformchannels/Channel.java
package com.microsoft.dotnet.platformchannels;

public abstract class Channel implements AutoCloseable {

    public Object sendMessageToDotNet(String messageId, Object[] args) throws Exception {

        if (managedHandler == null)
            throw new Exception("No MAUI Runtime registration for this Channel");

        return managedHandler.onChannelMessage(messageId, args);
    }

    public abstract Object handleMessageFromDotNet(String messageId, Object[] args);

    ChannelMessageHandler managedHandler;

    public void setManagedHandler(ChannelMessageHandler handler) {
        managedHandler = handler;
    }


    @Override
    public void close() throws Exception {
        managedHandler = null;
    }
}

File: Platforms/Apple/DotNetPlatformChannels/DotNetPlatformChannels/Channel.m
#import <objc/runtime.h>
#import "Channel.h"

@implementation Channel
- (id) sendMessageToDotNet: (NSString*)messageId withArgs:(NSArray*)args {
    return [[self managedHandler] onChannelMessage:messageId withArgs:args];
}

- (id<ChannelMessageHandler>)managedHandler {
    return objc_getAssociatedObject(self, @selector(managedHandler));
}

- (void)setManagedHandler:(id<ChannelMessageHandler>)handler {
    objc_setAssociatedObject(self, @selector(managedHandler), handler, OBJC_ASSOCIATION_RETAIN_NONATOMIC);
}

@end
